### ISSUE TITLE
CSS "hook" for left and top styles makes no sense

### ISSUE THREAD

**Issue by david-mark** [jquery/jquery#3448]
### Description ###

Nothing more pointless in browser scripting than trying to compute styles. The `document.defaultView.getComputedStyle` (incorrectly referenced as `window.getComputedStyle` in jQuery) method is a hideously complex black box that virtually never needs to be opened. In twenty years of browser scripting, I've managed to avoid it almost entirely (exception that comes to mind was an HTML editor).

Regardless, jQuery aspires to fix the whole thing and frequently gets into trouble. Consider this "hook" for the `top` and `left` styles:

```javascript
// Support: Safari <=7 - 9.1, Chrome <=37 - 49
// Add the top/left cssHooks using jQuery.fn.position
// Webkit bug: https://bugs.webkit.org/show_bug.cgi?id=29084
// Blink bug: https://bugs.chromium.org/p/chromium/issues/detail?id=589347
// getComputedStyle returns percent when specified for top/left/bottom/right;
// rather than make the css module depend on the offset module, just check for it here
jQuery.each( [ "top", "left" ], function( i, prop ) {
	jQuery.cssHooks[ prop ] = addGetHookIf( support.pixelPosition,
		function( elem, computed ) {
			if ( computed ) {
				computed = curCSS( elem, prop );

				// If curCSS returns percentage, fallback to offset
				return rnumnonpx.test( computed ) ?
					jQuery( elem ).position()[ prop ] + "px" :
					computed;
			}
		}
	);
} );
```

The cryptically named `support.pixelPosition` flag refers to a test result where a percentage-based `top` style fails to compute to pixel units. This is an issue that dates back to the bad old days of IE 8-, where `currentStyle` was frequently used by libraries as a slightly-off substitute for `getComputedStyle`; it didn't convert *any* units to pixels. jQuery certainly does this in older versions, so the problem should be familiar.

This is the part that attempts to fix the issue:

```javascript
// If curCSS returns percentage, fallback to offset
return rnumnonpx.test( computed ) ?
	jQuery( elem ).position()[ prop ] + "px" :
	computed;
```

It calls the `position` method instead, which attempts to figure the *offset* of the element from the document origin, which would only be equivalent to the computed `left` and `top` styles in only a small subset of cases.

For example, an element with a position style of `absolute` with all parent elements having position: `static` would work. On the other hand, an element with `fixed` position would fail *unless* the document's scroll position is its origin. Furthermore, an element with `relative` position would fail in virtually every case. And so on, this "hook" will simply cause browsers demonstrating this quirk to return nonsensical position values in numerous cases (not even considering problems related RTL documents).

For those who never bothered to read my "P is for Position" primer on the old cinsoft.net site (now retired) or the numerous Usenet articles I authored on this subject (dating back to at least 2008), I'll go over the general solution again.

In short:

1. Save the `left` or `top` style (e.g. `elem.style.left`).
2. Get the offset.
3. Set the `left` or `top` style to `'0'`.
4. Get the offset again.
5. Find the *difference* between the two offsets.
6. Put the old `left` or `top` style back.

And as the `offset` method is just a pointless wrapper for `getBoundingClientRect` (seems to exist only to handle calls passing disembodied elements or elements with a display style of `none`), would advise simply calling `getBoundingClientRect` and documenting that trying to compute position styles (or offsets) for such elements will result in undefined behavior (a recurring theme here). In addition, to let callers know they made a mistake at the earliest possible time, throw an exception in such cases (as opposed to returning zeros).

HTH

PS. What about `right` and `bottom` positions? The solution as it sits is only half-finished. Luckily I archived the related primer years ago:

https://github.com/david-mark/My-Library/blob/master/position.html

PPS. As this is an old issue that surely affects previous jQuery versions, am curious as to how the old versions are kept up to date when bugs like this are fixed.

### Link to test case ###

No test case needed. Logic is clearly incorrect.

**Comment by gibson042** [comment]
> It calls the `position` method instead, which attempts to figure the *offset* of the element from the document origin, which would only be equivalent to the computed `left` and `top` styles in only a small subset of cases.

See http://api.jquery.com/position/ (emphasis mine): "Get the current coordinates of the first element in the set of matched elements, _relative to the offset parent_."

> No test case needed. Logic is clearly incorrect.

This position which will not be tolerated. You are claiming the existence of a bug, and the onus is upon you to demonstrate it.

**Comment by david-mark** [comment]
> See http://api.jquery.com/position/ (emphasis mine): "Get the current coordinates of the first element in the set of matched elements, relative to the offset parent."

Granted, I must have had the `offset` method in mind. But regardless, that only invalidates a portion of the issues raised in this ticket.

> This position which will not be tolerated. You are claiming the existence of a bug, and the onus is upon you to demonstrate it.

When something is clearly wrong the code logic (e.g. the `position` function won't work for RTL documents either), then evidence gathering is strictly a secondary concern. As you often close tickets prematurely without sufficient discussion of the points raised (and as issues are not always bugs reproducible in presently supported browsers, but may indicate the potential for future issues), I can't see wasting my time on test cases until we are all on the same page and agree to address the contained issues.

We either understand what the code does or we don't. I made one mistake on this one, but there are other points that are still clearly valid. Now the ticket is closed, so glad I didn't bother getting too far into it.

For example, there's the issue of the missing `right` and `bottom` styles. There should be no test case needed for that as it's simply an omission. The code is there or it isn't. Quite sure I didn't miss that and not writing a test case to prove the code is missing.

**Comment by gibson042** [comment]
> Quite sure I didn't miss that and not writing a test case to prove the code is missing.

Then you're just making noise. I recommend you don't report bugs if you're not willing to provide test cases.

**Comment by david-mark** [comment]
> Then you're just making noise.

And how do you figure that? There are simply no "hooks" for right/bottom. All it would take is to search jQuery.js for "right" and "bottom". One line:

```javascript
var cssExpand = [ "Top", "Right", "Bottom", "Left" ];
```

And we both know that has nothing to do with the issue at hand. No test case needed. And as you have prematurely closed the ticket, I'm *certainly* not going to write one.

Furthermore, to address your previous objection:

> See http://api.jquery.com/position/ (emphasis mine): "Get the current coordinates of the first element in the set of matched elements, relative to the offset parent."

Shouldn't even need to look at the code to realize your inference that this fixes the central issue raised here is false. For example, what about elements with a `position` style of `relative`? How would finding the offset relative to the offset *parent* help with those?

Will handle a handful a few edge cases that work by coincidence, but should be obvious that the strategy will fail far more often than it works. Do we really need a test to prove this? Regardless, I'm definitely not writing any test cases for closed tickets.

Can't see the point in closing tickets before they have been discussed to the point where *everyone* understands the issues raised, as well as the proposed solutions. Is it an accounting thing?

Was told this was the best place to discuss proposed changes and I'm not seeing that at all. What I am seeing is a lot of confused, knee-jerk reactions and complaints that I'm just taking up time here. Easy solution to that. :)

**Comment by gibson042** [comment]
> Was told this was the best place to discuss proposed changes and I'm not seeing that at all.

Unlike some of the code reorganizations you've proposed, this ticket purports to describe a bug and calls for a behavior change. Nothing of the sort would be accepted without unit tests, and it's pointless to even start changing code without a demonstration of the problem.

If you're not willing to provide test cases, then your bug reports will almost certainly be closed so it makes no sense to file them. But if you do provide one, we will reopen this issue.

**Comment by david-mark** [comment]
> Unlike some of the code reorganizations you've proposed, this ticket purports to describe a bug and calls for a behavior change.

As well as adding code that clearly doesn't exist. All you have to do is look at the code in jQuery.js and see there are no hooks for `right` and `bottom` styles. As the code comment I quoted indicates, the bug we are trying to fix is not found exclusively in `top` and `left` styles. Shouldn't take a comment to realize that would be the case as elements can be positioned by percentage with `right` and `bottom` styles too.

As for the rest of it. Do we really need a test to determine that finding the offset from a positioned `parent` (if any) will not be equivalent to a `relative` position style (so it makes no sense to have that as the fallback?).

Would also have to dig up a browser that actually displays the quirks that the alleged fixes try to deal with. Much simpler to just agree that the logic in both these cases is wrong. It's *obviously* wrong and no tests could ever disprove that, no matter how many browsers and documents we try to test.

> Nothing of the sort would be accepted without unit tests

Unit tests should be added after we all agree on exactly what we are trying to fix and come up with a fix that demonstrably corrects the problems. Then the tests ensure that we have no regressions on these fixes in future.

And obviously the current unit tests are insufficient as they aren't testing any of the cited issues.

**Comment by dmethvin** [comment]
> Unit tests should be added after we all agree on exactly what we are trying to fix

Yeah, what *are* we trying to fix? Has a user reported a bug about this?

**Comment by david-mark** [comment]
> Yeah, what are we trying to fix? Has a user reported a bug about this?

See above. How much more explanation is necessary? To summarize:

1. Glaring omission of `right` and `bottom` style hooks. As if `left` and `top` are the only way to position elements.

2. Obvious mistake in using `position` method as a fallback for failed computing of `left` and `top` styles. For one example, it won't work for `position:relative`.

3. Another obvious mistake in thinking that `getBoundingClientRect` is appropriate to figure `left` and `top` *styles* in RTL documents (or parts of documents).

If you have questions about these issues, then *ask* for further explanation. Discussing issues is how we learn, which helps to avoid repeated mistakes in the future. And fixing issues now prevents them from being compounded in the future. Closing tickets without so much as a question is not productive. From your previous comments, I can only conclude that you are watching the bug tracker scoreboard and trying to make the "score" look more respectable.

Anyway, it's best to try to fix bugs *before* they get to the users. What sort of backwards quality control strategy waits until customers are already affected (and will have to "upgrade" a huge, complicated and monolithic DOM library) to fix the problem? Never mind that they may well pick up new bugs in the process as the code is in a constant state of flux.

And waiting around for a user to complain here is pointless. Most will first look for an "answer" on StackOverflow and get a dozen "try this!" responses. Some of them will involve other jQuery methods, others will bypass jQuery entirely by tacking on a subscript to the object and calling DOM methods. In any event, you'll remain blissfully ignorant to the problem.

Besides, there appears to be a perpetual backlog here and I count two people working on the 70 issues that are *currently* open. Wouldn't you try to avoid such a queue and go with a guess from StackOverflow too?

**Comment by david-mark** [comment]
See little evidence of teamwork here. If the person that reports the issue doesn't do all the work, then the ticket is simply closed without discussion so that nobody else will bother to pick it up.

Can't see the point of sweeping bugs under the rug. No matter how many are open at any given time, should be able to prioritize and allocate resources accordingly.

**Comment by dmethvin** [comment]
> See little evidence of teamwork here. If the person that reports the issue doesn't do all the work, then the ticket is simply closed without discussion so that nobody else will bother to pick it up.

We have to agree that there is work that needs to be done.  

> No matter how many are open at any given time, should be able to prioritize and allocate resources accordingly.

We're doing that, and I've been encouraging you to do the same. Focus on the bugs that the users of the library are reporting. Theoretical bugs or performance issues that nobody has encountered are not worth the time to triage or "fix".

**Comment by david-mark** [comment]
> We have to agree that there is work that needs to be done.

And, after all that, we still aren't in agreement on these issues?

> We're doing that, and I've been encouraging you to do the same. Focus on the bugs that the users of the library are reporting.

No, you are *closing* tickets instead of prioritizing them. That discourages others from working on them and the bugs just fall through the cracks. That's likely a big part of how the code has ended up in its present state.

> Theoretical bugs or performance issues that nobody has encountered are not worth the time to triage or "fix".

Theoretical bugs? Just because you don't understand them doesn't make them any less real. And you have no way to know that nobody has encountered them. I've explained that repeatedly as well.

And you don't have to take the time to "triage". Just don't close them until *somebody* takes the time to understand them.

**Comment by ctomczyk** [comment]
> Has a user reported a bug about this?

Why fixing bug is determined by the moment until the user report the bug? If somebody find the bug and it is not reported by any user does it means that the problem never occurred for any user? I am just trying to understand the process of workflow.

From my experience I see that the most of the users (including developers) will never report any bugs instead they will fix it "locally" as they need it quite fast and don't want to wait until specific owners/group of people will fix it.

**Comment by mgol** [comment]
>> Has a user reported a bug about this?
>
> Why fixing bug is determined by the moment until the user report the bug?

The biggest problem with this issue (& many others) is that it conflates many issues into one and is not clear about effects requested changes. It's really hard to follow such issues, as they promote digressions in the discussion. If someone reports a _separate_ issue which clearly requests _one_ API change with clear and _succint_ description, preferably with short code snippets showing the change then we'd consider.

> From my experience I see that the most of the users (including developers) will never report any bugs instead they will fix it "locally" as they need it quite fast and don't want to wait until specific owners/group of people will fix it.

That's true but if an issue affects many people, someone will usually report it in the end. We can't be second-guessing the users of the library.

**Comment by david-mark** [comment]
> The biggest problem with this issue (& many others) is that it conflates many issues into one and is not clear about effects requested changes.

No, it's all about that one "hook" related to `left` and `top` styles.

>  It's really hard to follow such issues, as they promote digressions in the discussion.

The jQuery code is a tangled mess in many places, so digressions will occur. Feel free to open additional issues as you see fit (as opposed to closing the door on all of them).

> If someone reports a separate issue which clearly requests one API change with clear and succint description, preferably with short code snippets showing the change then we'd consider.

Closing the ticket precludes any sort of teamwork on the issue. Few will bother to read it after it is closed.

> That's true but if an issue affects many people, someone will usually report it in the end.

In the end? :)

> We can't be second-guessing the users of the library.

What does that mean? We are the *coders* and therefore must be able to spot invalid logic before it makes its way to the users. At the very least, we must be proactive in fixing bugs that have already been made public. This has nothing to do with second-guessing (or any sort of guessing).

**Comment by gibson042** [comment]
@ctomczyk: If @david-mark or

**Comment by ahnzaz** [comment]
😲

2016. 12. 22. 오후 2:08에 "Richard Gibson" <notifications@github.com>님이 작성:

**Comment by david-mark** [comment]
> If @david-mark (or anyone else) provides a test case demonstrating the bug, we will reopen this issue.

So you still don't understand the issues? Not one of them? :(

**Comment by dmethvin** [comment]
A "bug report" that finishes with "No test case needed. Logic is clearly incorrect" is already starting off on the wrong foot. *Start* with a test case that clearly fails, then we can talk about whether/why the code is wrong and how to fix it. The test case will end up in unit tests and we will all be happy.

**Comment by david-mark** [comment]
> A "bug report" that finishes with "No test case needed. Logic is clearly incorrect" is already starting off on the wrong foot.

Not really as we should be able to agree that the logic is wrong without seeing it fail in a test case (a recurring theme). And never mind starting off; it's long past starting off and the ticket is closed. The only solution to that is to re-open it. Then maybe *somebody* will work on it. As long as it stays closed, then *nobody* is going to work on it (particularly not me). Closing the ticket tells everyone who may be interested in helping to move on (nothing to see here).

> Start with a test case that clearly fails, then we can talk about whether/why the code is wrong and how to fix it.

In this particular case, what makes you think that any "supported" browsers even run that BS "hook" code? I already know that if the test fails outside the most recent versions of a handful of browsers then it will be rejected as an irrelevant result.

So the "hook" that never even runs today will remain in the code because it doesn't fail any tests in "current" browsers. Do you follow me here? This is why you can't rely on test results for everything; have to actually understand what the code is doing to analyze test results.

The solution I propose for this is to remove whole mess. Then run the *current* unit tests in whatever narrow range of browsers we consider "supported" and see what happens. But again, nobody is going to do anything with a closed ticket. ;)
